### ISSUE TITLE
Do not use HEREDOC or NOWDOC syntax in your plugins

### ISSUE THREAD

**Issue by ernilambar** [10up/plugin-check#792]
### Do not use HEREDOC or NOWDOC syntax in your plugins

> While both are totally valid, and in many ways desirable features of PHP that allow you to output content, it comes with a cost that is too high for most plugins.

> The primary issue is that most (if not all) codesniffers won't detect lack of escaping in code when you use HEREDOC or NOWDOC. While there are ways around this they have the end result of dashing all that readability to the rubbish pile and leaving you with a jumbled mess that won't properly be scanned.

> We feel the risk here is much higher than the benefits, which is why we don't permit their use.

We could use this rule - https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Sniffs/PHP/HeredocSniff.php

**Comment by davidperezgar** [comment]
Ok!

**Comment by westonruter** [comment]
I'm curious about the rationale here. I can understand the issue with HEREDOC since you can't add escaping functions in the output for interpolated variables, but this isn't the case for NOWDOC. 

For example, in https://github.com/WordPress/performance/pull/2144, I want to do this:

```php
$js = <<<'JS'
	const authRadios = /** @type {NodeListOf<HTMLInputElement>} */ ( document.querySelectorAll( "input[name='plsr_speculation_rules[authentication]']" ) );
	const noticeDiv = document.getElementById( "plsr-auth-notice" );
	if ( authRadios.length && noticeDiv ) {
		for ( const authRadio of authRadios ) {
			authRadio.addEventListener( "change", () => {
				const isLoggedOut = ( authRadio.value === "logged_out" );
				noticeDiv.classList.toggle( "notice-info", isLoggedOut )
				noticeDiv.classList.toggle( "notice-warning", ! isLoggedOut )
			} );
		}
	}
JS;
wp_print_inline_script_tag( $js, array( 'type' => 'module' ) );
```

But the PCP ruleset is flagging this. What is the risk here?

**Comment by westonruter** [comment]
Note: There are 52 files in WordPress core using heredoc/nowdoc: https://github.com/search?q=repo%3AWordPress%2Fwordpress-develop+%2F%3C%3C%3C%27%3F%5BA-Z%5D%2B%2F+path%3A*.php&type=code

**Comment by westonruter** [comment]
I've opened https://github.com/WordPress/plugin-check/issues/1036 to discuss this.
